user-guide/Advanced_Modules/Dashboards_and_Low_Code_Apps/GQI/Extensions/Ad_hoc_data/Ad_hoc_Tutorials_GQIDMS.md

```diff
@@ -0,0 +1,121 @@
+---
+uid: Ad_hoc_Tutorials_GQIDMS
+---
+
+# Building a GQI data source that retrieves data from a DMS
+
+In this tutorial, you will learn how you can create a GQI data source that retrieves the active client connections from your DMS.
+
+Expected duration: 15 minutes.
+
+> [!NOTE]
+> This tutorial uses DataMiner version 10.3.4/10.4.0 onwards.
+
+## Prerequisites
+
+- [Visual Studio](https://visualstudio.microsoft.com/downloads/) with [DataMiner Integration Studio](xref:Installing_and_configuring_DataMiner_Integration_Studio)
+
+## Overview
+
+- [Step 1: Create the ad hoc data source](#step-1-create-the-ad-hoc-data-source)
+- [Step 2: Fetch the client connections from the DMS](#step-2-fetch-the-client-connections-from-the-dms)
+- [Step 3: Parse the responses into a GQIPage](#step-3-parse-the-responses-into-a-gqipage)
+
+## Step 1: Create the ad hoc data source
+
+1. Create a new class that implements the [IGQIDatasource](xref:GQI_IGQIDataSource) interface.
+
+> [!NOTE]
+> If certain types cannot be found in the file, verify if the *Skyline.DataMiner.Dev.Automation* NuGet package has the correct version. Go to *Tools* > *NuGet Package Manager* > *Manage NuGet Packages for Solution*. Select *Skyline.DataMiner.Dev.Automation*, and verify whether the version installed for the current project is at least *10.3.4*.
+
+1. Implement the GetColumns function to define the columns that the data source will contain. Our data source will include two columns: `Client` (of type `String`) and `Connection Time` (of type `DateTime`).
```

Comment:
In C# member functions in classes and interfaces are always called 'methods' instead. Also use consistent formatting for code elements vs. names.

Could also link each method reference to its paragraph in the docs.

```suggestion
1. Implement the `GetColumns` method to define the columns of the data source. Our data source will include two columns: a *Client* column of  type `String` and a *Connection Time* column of type `DateTime`.
```